AudioDeviceExample/AudioDevices/ExampleAudioEngineDevice.m

```diff
@@ -0,0 +1,741 @@
+     * Configure the RemoteIO audio unit. Our rendering format attempts to match what AVAudioSession requires to
+     * prevent any additional format conversions after the media engine has mixed our playout audio.
+     */
+    AudioStreamBasicDescription streamDescription = self.renderingFormat.streamDescription;
+
+    UInt32 enableOutput = 1;
+    status = AudioUnitSetProperty(_audioUnit, kAudioOutputUnitProperty_EnableIO,
+                                  kAudioUnitScope_Output, kOutputBus,
+                                  &enableOutput, sizeof(enableOutput));
+    if (status != 0) {
+        NSLog(@"Could not enable out bus!");
+        AudioComponentInstanceDispose(_audioUnit);
+        _audioUnit = NULL;
+        return NO;
+    }
+
+    status = AudioUnitSetProperty(_audioUnit, kAudioUnitProperty_StreamFormat,
+                                  kAudioUnitScope_Output, kInputBus,
+                                  &streamDescription, sizeof(streamDescription));
+    if (status != 0) {
+        NSLog(@"Could not enable output bus!");
+        return NO;
+    }
+
+
+
+    status = AudioUnitSetProperty(_audioUnit, kAudioUnitProperty_StreamFormat,
+                                  kAudioUnitScope_Input, kOutputBus,
+                                  &streamDescription, sizeof(streamDescription));
+    if (status != 0) {
+        NSLog(@"Could not enable output bus!");
+        return NO;
+    }
+
+    // Enable the microphone input
+    UInt32 enableInput = 1;
+    status = AudioUnitSetProperty(_audioUnit, kAudioOutputUnitProperty_EnableIO,
+                                  kAudioUnitScope_Input, kInputBus, &enableInput,
+                                  sizeof(enableInput));
+
+    if (status != 0) {
+        NSLog(@"Could not enable input bus!");
+        AudioComponentInstanceDispose(_audioUnit);
+        _audioUnit = NULL;
+        return NO;
+    }
+
+    // Setup the rendering callback.
+    AURenderCallbackStruct renderCallback;
+    renderCallback.inputProc = ExampleCoreAudioDevicePlayoutCallback;
+    renderCallback.inputProcRefCon = (void *)(renderContext);
+    status = AudioUnitSetProperty(_audioUnit, kAudioUnitProperty_SetRenderCallback,
+                                  kAudioUnitScope_Output, kOutputBus, &renderCallback,
+                                  sizeof(renderCallback));
+    if (status != 0) {
+        NSLog(@"Could not set rendering callback!");
+        AudioComponentInstanceDispose(_audioUnit);
+        _audioUnit = NULL;
+        return NO;
+    }
+
+    // Setup the capturing callback.
+    AURenderCallbackStruct captureCallback;
+    captureCallback.inputProc = ExampleCoreAudioDeviceCaptureCallback;
+    captureCallback.inputProcRefCon = (void *)(self.capturingContext);
+    status = AudioUnitSetProperty(_audioUnit, kAudioOutputUnitProperty_SetInputCallback,
+                                  kAudioUnitScope_Input, kInputBus, &captureCallback,
+                                  sizeof(captureCallback));
```

Comment:
Since we are pre-allocating our own AudioBufferList for input should we tell the AudioUnit not to worry about allocating its own buffer? (kAudioUnitProperty_ShouldAllocateBuffer)